### PR TITLE
Add DIY nRF82540 PCA10059 4.2inch NiceRF 868

### PR DESCRIPTION
**Pull request by markbirss**
This requires

https://github.com/meshtastic/Meshtastic-protobufs/pull/137

Kindly review and comment

### CODE REVIEW

File: variants/Dongle_nRF52840-pca10059-v1/variant.h [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+#define LED_STATE_ON 0              // State when LED is litted
+
+/*
+ * Buttons
+ */
+
+#ifdef PCA10059
+#define PIN_BUTTON1 (32 + 6)        // BTN_DN           P1.06 Built in button
+#endif
```
Comment (caveman99) [inline variants/Dongle_nRF52840-pca10059-v1/variant.h:71]: Is the guard neccessary here? we're board specific anyway in this file.

File: variants/Dongle_nRF52840-pca10059-v1/variant.h [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+#define SX126X_E22
+
+#define PIN_GPS_EN     (-1)
+#define PIN_GPS_PPS    (-1)           // Pulse per second input from the GPS
+
+#ifdef PCA10059
+#define GPS_RX_PIN PIN_SERIAL1_RX
+#define GPS_TX_PIN PIN_SERIAL1_TX
+#endif
+
```
Comment (caveman99) [inline variants/Dongle_nRF52840-pca10059-v1/variant.h:155]: See above ... Is the guard neccessary?

**Review by caveman99 — CHANGES_REQUESTED** [review]
Looking fabulous, some small remarks. Nagging on a high level.

File: src/graphics/EInkDisplay2.cpp [hunk 3]
```diff
@@ -71,8 +81,8 @@ bool EInkDisplay::forceDisplay(uint32_t msecLimit)
 
         // FIXME - only draw bits have changed (use backbuf similar to the other displays)
         // tft.drawBitmap(0, 0, buffer, 128, 64, TFT_YELLOW, TFT_BLACK);
-        for (uint64_t y = 0; y < displayHeight; y++) {
-            for (uint64_t x = 0; x < displayWidth; x++) {
+        for (uint32_t y = 0; y < displayHeight; y++) {
+            for (uint32_t x = 0; x < displayWidth; x++) {
```
[the remark this answers is not in the document]
Comment (markbirss) [reply src/graphics/EInkDisplay2.cpp:85]: Yes, i first had to change this when adding the RAK14000 e-paper and found that they can be made smaller. (uint8 is too small)

File: src/graphics/EInkDisplay2.cpp [hunk 4]
```diff
@@ -85,6 +95,8 @@ bool EInkDisplay::forceDisplay(uint32_t msecLimit)
         DEBUG_MSG("Updating T-ECHO E-Paper... ");
         #elif defined(RAK4630)
         DEBUG_MSG("Updating RAK4361_5005 E-Paper... ");
+        #elif defined(PCA10059)
+        DEBUG_MSG("Updating PCA10059 E-Paper... ");
```
[the remark this answers is not in the document]
Comment (markbirss) [reply src/graphics/EInkDisplay2.cpp:99]: Sure, can be done, as there will be 3 boards with e-paper

File: variants/Dongle_nRF52840-pca10059-v1/variant.h (hunk 1 above)
Comment (markbirss) [reply variants/Dongle_nRF52840-pca10059-v1/variant.h:71]: I will remove it

File: variants/Dongle_nRF52840-pca10059-v1/variant.h (hunk 2 above)
Comment (markbirss) [reply variants/Dongle_nRF52840-pca10059-v1/variant.h:155]: I will have a look, shortly and update

**Review by caveman99 — APPROVED** [review]

**Review by markbirss — COMMENTED** [review]
Kindly review as I made those changes suggested

**Comment by caveman99** [comment]
For the conflict: Safest bet is to make a pull request from meshtastic-Master upstream to your fork and incorporate it there, then push it to this PR.

**Comment by markbirss** [comment]
@thebentern hi, I see the actual check and build of fw pca10059_diy_eink is still missing from above ?

**Comment by thebentern** [comment]
Oh, we missed the github action. I need to remove build-all and check-all. They're no longer used

**Comment by geeksville** [comment]
This pull request has been mentioned on **Meshtastic**. There might be relevant details there:

https://meshtastic.discourse.group/t/wip-diy-nrf82540-nicerf-868-4-2-inch-epaper/5552/3

**Pull request merged**